coord: enable consistency topics for JSON-encoded Kafka sinks

JLDLaughlin opened the pull request:
This PR allows JSON-encoded Kafka sinks to use consistency topics (and the `reuse_topic` flag).

Next step: docs for JSON sinks!

philip-stoev commented on src/sql/src/plan/statement/ddl.rs:
```diff
@@ -1316,17 +1310,86 @@ fn kafka_sink_builder(
+                )
+            }
+            Some(other) => unsupported!(format!("consistency format {}", &other)),
+            None => unsupported!("consistency topic without format"),
```
I think this message should somehow say that  `CONSISTENCY FORMAT` is expected, as the statement already has a `FORMAT` in it.

To make it even easier for the user, can the `FORMAT` be used as `CONSISTENCY FORMAT` unless overridden?

JLDLaughlin replied:
Great suggestion, thank you! I've done this:
> To make it even easier for the user, can the FORMAT be used as CONSISTENCY FORMAT unless overridden?
And added [a test here](https://github.com/MaterializeInc/materialize/pull/7608/commits/d90aad3574122cffdb3326d1ebb4af3dbd88c266#diff-37787161b8fbbc6abdba42d5590532844a08e9a73c4f5304a3228375af7b75bdR230).

philip-stoev commented on test/testdrive/kafka-exactly-once-sink.td:
```diff
@@ -199,3 +199,55 @@ $ kafka-ingest format=avro topic=compaction-test-input-consistency timestamp=1 s
+> CREATE SINK double_avro FROM unnamed_cols
+  INTO KAFKA BROKER '${testdrive.kafka-addr}' TOPIC 'double-avro'
+    CONSISTENCY TOPIC 'consistency-double-avro' CONSISTENCY FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY '${testdrive.schema-registry-url}'
+  FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY '${testdrive.schema-registry-url}'
```
I guess this is part of the previous PR that I slept through, but can this syntax be made more compact? e.g. assume the same schema registry and format will be used for the consistency topic. Right now specifying ```CONFLUENT SCHEMA REGISTRY``` twice is mandatory:

```
CREATE SINK double_avro FROM unnamed_cols INTO KAFKA BROKER 'localhost:9092' TOPIC 'double-avro'   CONSISTENCY TOPIC 'consistency-double-avro' CONSISTENCY FORMAT AVRO FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY 'http://localhost:8081/': Expected USING, found FORMAT
```

JLDLaughlin replied:
I think I fixed this by implementing your suggestion (that the user should be able to elide `CONSISTENCY FORMAT` if it's the same as `FORMAT`. Let me know what you think!

philip-stoev commented:
Actually using the feature as described in the title of this PR unfortunately results in an immediate panic:

```
> CREATE VIEW unnamed_cols AS SELECT 1, 2 AS b, 3;

> CREATE SINK json_sink FROM unnamed_cols
  INTO KAFKA BROKER '${testdrive.kafka-addr}' TOPIC 'json-sink'
  CONSISTENCY TOPIC 'consistency-json-sink' CONSISTENCY FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY '${testdrive.schema-registry-url}'
  FORMAT JSON;
```

```
2021-07-30T06:51:05.667678Z ERROR panic: itertools: .zip_eq() reached end of one iterator before the other
thread: timely:work-0
location: /home/pstoev/.cargo/registry/src/github.com-1ecc6299db9ec823/itertools-0.10.1/src/zip_eq_impl.rs:48:13
version: 0.8.4-dev (5474b1417a05fd30c26155546e63c7f9e34db9e3)
backtrace:
```

This does not happen unless ```CONSISTENCY TOPIC``` is specified.

aljoscha commented:
Could you add a test that verifies JSON data output with `reuse_topic=true` uses default Avro consistency output?

Overall, I think this looks very good already? Modulo the things Philip discovered, and which should be covered by adding more tests as well.

JLDLaughlin commented:
@philip-stoev -- thanks for catching that! I've fixed the bug (I wasn't correctly updating the `value_desc` to send transaction ids) and I've added a few tests. Please let me know if there are other tests you think I should add!

@aljoscha -- I've added tests that combine `reuse_topic` and `FORMAT JSON`!

philip-stoev reviewed (approved):
Thank you for the extra tests and for incorporating my suggestions. I could not think of anything else to test at this time.

aljoscha commented on test/testdrive/kafka-exactly-once-sink.td:
```diff
@@ -228,24 +228,93 @@ $ kafka-verify format=avro sink=materialize.public.rt_binding_consistency_test_s
+  INTO KAFKA BROKER '${testdrive.kafka-addr}' TOPIC 'json-reuse-topic'
+    WITH (reuse_topic=true)
+  FORMAT JSON
+CONSISTENCY FORMAT JSON not yet supported
```
This one is a bit confusing, because the user didn't specify `CONSISTENCY FORMAT JSON` themselves. But I think the alternative would be to default to Avro consistency topic when using a JSON sink with `reuse_topic`. Which, I'm also not sure that's good.

Boy, this stuff is getting complicated... 😅 

The solution is probably future support for a JSON consistency topic. :trollface:

JLDLaughlin replied:
Yeah, I agree!

The cleanest solution might be to not default to anything. While this is super explicit, it's not the most user-friendly experience. (This is what I originally did, but Philip pointed out that a default might be helpful).

It will get cleaner once we support a JSON consistency topic for sure!

aljoscha reviewed (approved):
This looks very good!

I had a nit, but I don't think we need to solve this now.

aljoscha commented:
There's still a test failure, though.
